equinox-shipping/Domain/FinalizationTransaction.fs

```diff
@@ -0,0 +1,91 @@
+module FinalizationTransaction
+
+module Events =
+
+    let [<Literal>] CategoryId = "FinalizationTransaction"
+    let streamName clientId = FsCodec.StreamName.create CategoryId clientId
+
+    type Event =
+        | FinalizationRequested of {| containerId : string; shipmentIds : string[] |}
+
+        | AssignmentCompleted   of {| containerId : string; shipmentIds : string[] |}
+        | FinalizationCompleted
+
+        | RevertRequested       of {| shipmentIds : string[] |}
+        | FinalizationFailed
+        interface TypeShape.UnionContract.IUnionContract
+
+    let codec = FsCodec.NewtonsoftJson.Codec.Create<Event>()
+
+type Action =
+    | AssignShipments   of containerId: string * shipmentIds: string[]
+    | FinalizeContainer of containerId: string * shipmentIds: string[]
+    // Reverts the assignment of the container for the shipments provided.
+    | RevertAssignment  of shipmentIds: string[]
+    | Finish of success: bool
+
+module Fold =
+
+    type State =
+        | Initial
+        | Running   of RunningState
+        | Completed of bool
+    and RunningState =
+        | Assigning of containerId: string * shipmentIds: string[]
+        | Assigned  of containerId: string * shipmentIds: string[]
+        | Reverting of shipmentIds: string[]
+
+    let initial: State = Initial
+
+    let evolve (_: State) (event: Events.Event): State =
+        match event with
+        | Events.FinalizationRequested event -> Running (Assigning (event.containerId, event.shipmentIds))
+
+        | Events.AssignmentCompleted   event -> Running (Assigned  (event.containerId, event.shipmentIds))
+        | Events.FinalizationCompleted       -> Completed true
+
+        | Events.RevertRequested       event -> Running (Reverting event.shipmentIds)
+        | Events.FinalizationFailed          -> Completed false
+
+    let nextAction (state: State): Action =
+        match state with
+        | Initial -> failwith "Cannot interpret Initial state"
+        | Running (Assigning (containerId, shipmentIds)) -> Action.AssignShipments   (containerId, shipmentIds)
+        | Running (Assigned  (containerId, shipmentIds)) -> Action.FinalizeContainer (containerId, shipmentIds)
+        | Running (Reverting shipmentIds)                -> Action.RevertAssignment  shipmentIds
+        | Completed result                               -> Action.Finish result
+
+    let fold: State -> Events.Event seq -> State =
+        Seq.fold evolve
+
+    let filterValidTransition (event: Events.Event) (state: State) =
```

Comment:
filter -> choose ? 
space before `:`

Reply:
> filter -> choose ?
What do you mean?

Reply:
Normally I'd expect a `filter` op to be doing a bool predicate; this one is a `chooser` for me

Reply:
Ah right that makes sense - I've renamed. Thanks!